src/seat/touch.rs

```diff
@@ -0,0 +1,184 @@
+use std::sync::Mutex;
+
+use wayland_client::protocol::wl_surface::WlSurface;
+use wayland_client::protocol::wl_touch::{Event as TouchEvent, WlTouch};
+use wayland_client::{
+    ConnectionHandle, DelegateDispatch, DelegateDispatchBase, Dispatch, QueueHandle,
+};
+
+use crate::seat::{SeatHandler, SeatState};
+
+#[derive(Debug, Default)]
+pub struct TouchData {
+    inner: Mutex<TouchDataInner>,
+}
+
+#[derive(Debug, Default)]
+pub(crate) struct TouchDataInner {
+    events: Vec<TouchEvent>,
+}
+
+#[macro_export]
+macro_rules! delegate_touch {
+    ($ty: ty) => {
+        $crate::reexports::client::delegate_dispatch!($ty:
+            [
+                $crate::reexports::client::protocol::wl_touch::WlTouch
+            ] => $crate::seat::SeatState
+        );
+    };
+}
+
+pub trait TouchHandler: SeatHandler + Sized {
+    /// New touch point.
+    ///
+    /// Indicates a new touch point has appeared on the surface, starting a touch sequence. The ID
+    /// associated with this event identifies this touch point for devices with multi-touch and
+    /// will be referenced in future events.
+    ///
+    /// The associated touch ID ceases to be valid after the touch up event with the associated ID
+    /// and may be reused for other touch points after that.
+    #[allow(clippy::too_many_arguments)]
+    fn down(
+        &mut self,
+        conn: &mut ConnectionHandle,
+        qh: &QueueHandle<Self>,
+        touch: &WlTouch,
+        serial: u32,
+        time: u32,
+        surface: WlSurface,
+        id: i32,
+        x: f64,
+        y: f64,
+    );
```

Comment:
I kind of agree with clippy here. A non-exhaustive struct should satisfy this while allowing future extension (if it happens) to occur.

Reply:
I've wanted to follow the behavior of the other handlers here and largely mimic the events and break everything out.

Starting to bundle things into structs for one API but not the others seems very inconsistent to me? Especially with things like `serial` having it as a parameter in one and a subparameter of some random parameter in another would be hugely irritating.

Reply:
I think I agree to keep consistency with the rest of the API, introducing a struct for this event only would be too confusing.

I guess an alternative option would be to have a `TouchPointEvent` enum that fuses `down`, `up`, `motion`, `shape` and `orientation` into a single `fn (&mut self, &mut ConnectionHandle, &QueueHandle<Self>, WlTouch, TourchPointEvent)` method.

Reply:
I don't think a single handler is the way to go tbh. Because that way you're splitting things up, then you match them in SCTK, just to combine them again and have the client match it again. That just seems unfortunate for downstream and at that point causes more harm than it does good.

There are some minor things one can do, combining x and y to a tuple is done in other APIs for example and iirc that might remove one of the warnings. Other than that it's rough though with touch down itself, there's lots to unpack there wrt information contained in it.

Besides the clippy ignore, which I don't really like myself, I think the best path would probably just be to keep frequently used information flat and obstruct infrequently used information. Though that highly depends on the usecase. One person might think `time` is mostly useless while another treats every `id` the same anyway.

Reply:
I've removed one of the clippy annotations, combining coordinates similar to what is done in `pointer.rs`. But I don't think there's a great way to do that for the `down` function without transforming all other functions too and creating something inconsistent to the other APIs.

So I've just left the annotation on there for now. It's certainly not ideal but I think it's okay.